_LEARNING-CIRCLE-24 PROJECT_

<h2> Learning-Circle-24 </h2> <p>  Contributors </p>



# **Crime in Chicago**
This is a dataset
The data dictionary can be found in [Chicago Crime Portal](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-Present/ijzp-q8t2/about_data)

In this notebook, we are going to explore more about crime in Chicago and try to answer few questions:

How has crime in Chicago changed across years? What year was really the bloodiest year in two decades?
Are some types of crimes more likely to happen in specific locations or specific time of the day or specific day of the week than other types of crimes?

Analyzing the Chicago crime dataset involves examining various aspects of the data to understand patterns, trends, and characteristics of criminal activities in Chicago. We perform analysis using Python, pandas and visualization libraries using the following steps:

1. Load the Dataset: Load the Chicago crime dataset into a DataFrame.
2. Explore the Dataset: Examine the structure, columns, and basic statistics of the dataset.
3. Data Cleaning: We dropped redundant columns and handle missing values.
4. Feature Engineering: Create new column for Months, Day, Season
5. EDA: Exploratory Data Analysis
    - How has crime in Chicago changed across years? Was 2023 really the bloodiest year in two decades?
    - Are some types of crimes more likely to happen in specific locations or specific time of the day or specific day of the         week than other types of crimes?
    - In which month did the lowest motor vehicle thefts occur?
    - How prominant are certain types of various crimes?
    - How do rates of crime change of time? For instance, are there any seasonal effects or changes depending on the time of         day?
    - Where do these crimes occur? Are some areas more safe than others?
    - 
6. Frequency distribution of Crimes: show how the frequency of crime is distributed within Months, Days, Seasons.
7. Visualize Data Distribution: Visualize the distribution of variables using histograms, box plots, or density plots.
8. Investigate Relationships: Look for relationships between variables using scatter plots, correlation matrices, or pair plots
9. Investigate Crime Types: Analyze the distribution of different types of crimes.
10. Examine Crime Locations: Investigate the distribution of crimes across different locations.
11. Identify Hotspots: Identify areas with high crime rates.
12. Detect Seasonal Patterns: Look for seasonal patterns or variations in crime rates.
13. Compare Crime Rates: Compare crime rates across different categories such as day of the week, time of day, and month.
14. Summarize and Visualize Insights: Summarize and Visualize the key insights and findings from the analysis using plots and charts.

In [2]:
# importing required libraries
import pandas as pd
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
# Set display option 
pd.set_option('display.max.columns', None)

# 1. Loading the dataset

In [4]:
# Load the dataset
chicago_crime_df= pd.read_csv('../Data-Science-Projects/dataset/crime_data_chicago.csv')

# 2. Explore the dataset

In [5]:
#checking the dimension of the dataset
chicago_crime_df.shape

(2278726, 23)

In [6]:
# Display the first five rows of the dataset
chicago_crime_df.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [7]:
# check the first five observations
chicago_crime_df.head()

Unnamed: 0        ID Case Number                    Date  \
0           0   6407111    HP485721  07/26/2008 02:30:00 PM   
1           1  11398199    JB372830  07/31/2018 10:57:00 AM   
2           2   5488785    HN308568  04/27/2007 10:30:00 AM   
3           3  11389116    JB361368  07/23/2018 08:55:00 AM   
4           4  12420431    JE297624  07/11/2021 06:40:00 AM   

                  Block  IUCR       Primary Type               Description  \
0  085XX S MUSKEGON AVE  1320    CRIMINAL DAMAGE                TO VEHICLE   
1     092XX S ELLIS AVE  143C  WEAPONS VIOLATION  UNLAWFUL POSS AMMUNITION   
2     062XX N TRIPP AVE  0610           BURGLARY            FORCIBLE ENTRY   
3    0000X N KEELER AVE  0560            ASSAULT                    SIMPLE   
4   016XX W HARRISON ST  051A            ASSAULT      AGGRAVATED - HANDGUN   

                     Location Description  Arrest  Domestic  Beat  District  \
0                                  STREET   False     False   423       4.0   
1                               POOL ROOM    True     False   413       4.0   
2                               RESIDENCE    True     False  1711      17.0   
3            NURSING HOME/RETIREMENT HOME   False     False  1115      11.0   
4  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231      12.0   

   Ward  Community Area FBI Code  X Coordinate  Y Coordinate  Year  \
0  10.0            46.0       14     1196638.0     1848800.0  2008   
1   8.0            47.0       15     1184499.0     1843935.0  2018   
2  39.0            12.0       05     1146911.0     1941022.0  2007   
3  28.0            26.0      08A     1148388.0     1899882.0  2018   
4  27.0            28.0      04A     1165430.0     1897441.0  2021   

               Updated On   Latitude  Longitude                       Location  
0  02/28/2018 03:56:25 PM  41.739980 -87.555120  (41.739979622, -87.555120042)  
1  08/07/2018 04:02:59 PM  41.726922 -87.599747  (41.726922145, -87.599746995)  
2  02/28/2018 03:56:25 PM  41.994138 -87.734959  (41.994137622, -87.734959049)  
3  07/30/2018 03:52:24 PM  41.881217 -87.730590  (41.881217483, -87.730589961)  
4  07/18/2021 04:56:02 PM  41.874174 -87.668082  (41.874173691, -87.668082118)

In [8]:
 # Display column information and data types
chicago_crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2278726 entries, 0 to 2278725
Data columns (total 23 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Unnamed: 0            int64  
 1   ID                    int64  
 2   Case Number           object 
 3   Date                  object 
 4   Block                 object 
 5   IUCR                  object 
 6   Primary Type          object 
 7   Description           object 
 8   Location Description  object 
 9   Arrest                bool   
 10  Domestic              bool   
 11  Beat                  int64  
 12  District              float64
 13  Ward                  float64
 14  Community Area        float64
 15  FBI Code              object 
 16  X Coordinate          float64
 17  Y Coordinate          float64
 18  Year                  int64  
 19  Updated On            object 
 20  Latitude              float64
 21  Longitude             float64
 22  Location              object 
dtypes: bool

In [9]:
# checking missing value count
chicago_crime_df.isnull().sum()

Unnamed: 0                   0
ID                           0
Case Number                  1
Date                         0
Block                        0
IUCR                         0
Primary Type                 0
Description                  0
Location Description      2877
Arrest                       0
Domestic                     0
Beat                         0
District                    12
Ward                    184695
Community Area          184267
FBI Code                     0
X Coordinate             23985
Y Coordinate             23985
Year                         0
Updated On                   0
Latitude                 23985
Longitude                23985
Location                 23985
dtype: int64

In [10]:
chicago_crime_df.describe()

Unnamed: 0            ID          Beat      District          Ward  \
count  2.278726e+06  2.278726e+06  2.278726e+06  2.278714e+06  2.094031e+06   
mean   1.139362e+06  6.882068e+06  1.186442e+03  1.129072e+01  2.272764e+01   
std    6.578117e+05  3.419168e+06  7.026836e+02  6.946692e+00  1.383464e+01   
min    0.000000e+00  6.370000e+02  1.110000e+02  1.000000e+00  1.000000e+00   
25%    5.696812e+05  3.716076e+06  6.210000e+02  6.000000e+00  1.000000e+01   
50%    1.139362e+06  6.885990e+06  1.034000e+03  1.000000e+01  2.300000e+01   
75%    1.709044e+06  9.887568e+06  1.731000e+03  1.700000e+01  3.400000e+01   
max    2.278725e+06  1.278199e+07  2.535000e+03  3.100000e+01  5.000000e+01   

       Community Area  X Coordinate  Y Coordinate          Year      Latitude  \
count    2.094459e+06  2.254741e+06  2.254741e+06  2.278726e+06  2.254741e+06   
mean     3.752140e+01  1.164569e+06  1.885747e+06  2.009638e+03  4.184209e+01   
std      2.153282e+01  1.673955e+04  3.209855e+04  6.019724e+00  8.830434e-02   
min      0.000000e+00  0.000000e+00  0.000000e+00  2.001000e+03  3.661945e+01   
25%      2.300000e+01  1.152948e+06  1.859053e+06  2.004000e+03  4.176866e+01   
50%      3.200000e+01  1.166060e+06  1.890673e+06  2.009000e+03  4.185578e+01   
75%      5.700000e+01  1.176365e+06  1.909219e+06  2.014000e+03  4.190668e+01   
max      7.700000e+01  1.205119e+06  1.951622e+06  2.022000e+03  4.202291e+01   

          Longitude  
count  2.254741e+06  
mean  -8.767161e+01  
std    6.073538e-02  
min   -9.168657e+01  
25%   -8.771379e+01  
50%   -8.766597e+01  
75%   -8.762823e+01  
max   -8.752453e+01

# 3. Data Preprocessing

## Data Cleaning

Here are some steps that we would carry out in this phase

1. Convert data types to the appropriate formats
2. Remove redundant or unnecessary columns
3. Handle missing values

__Reasons for dropping the below columns__
- 'Unnamed: 0': Its just an index for individual row
- 'ID': Serial identifier for the record which is redundant due to the presence of a case number for each case.
- 'Location': This is where the incident occurred in a format that allows for creation of maps and other geographic operation. It's the combination of both latitude and longitude feature of this dataset thus, not required.
- 'Updated On': Holds the date and time the record was last updated.Which is not required for our analysis.
- 'X Coordinate': The x coordinate is a shifted location from the actual location for partial redaction but falls on the same block. Since we have a longitude/latitude and the block where the crime happened, we can neglect this feature.
- 'Y Coordinate': The y coordinate is a shifted location from the actual location for partial redaction but falls on the same block. Since we have a longitude/latitude and the block where the crime happened, we can neglect this feature.
- 'Community Area': This indicates the community area where the incident occurred. Which is not neccessary for our analysis.
- 'IUCR': This is the Illinois Unifrom Crime Reporting code. This is directly linked to the Primary Type and Description, it provides information about the primary type and description of crime. Since primary type and description features already exist, it becomes a redundant column
- Description: It's simply a subcategory of the primary type(which describes the type of crime committed). this can be omitted
- Block: Since we have the latitude and longitude, this column is redundant.
- Community Area: This is not needed in our analysis and the presence of a police Beat makes this redundant. 
- FBI Code: Removed because it is not necessary for our analysis as we are unable to de-code these codes.

__1. convert inappropriate data types to the correct one__

In [11]:
# first we convert the date column to datetime data type.

# chicago_crime_df['Date'] = pd.to_datetime(chicago_crime_df['Date'],format = "%m/%d/%Y %I:%M:%S %p")

In [12]:
# convert the district column to a string; from the keynote, Chicago police has only 22 district and the data type is plain text

chicago_crime_df['District'] = pd.Series(chicago_crime_df['District'], dtype='string')

__2. Use a funtion to Drop redundant columns and/or columns that are not necessary for our analysis (a redundant column is one that gives infomation that another column already explains: eg ID and Case number)__

In [13]:
chicago_crime_df.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [14]:
# Dropping redundant columns
def drop_redundant_columns(df):
    redundant_columns = ['Unnamed: 0','ID', 'Location','Updated On', 'FBI Code', 'Block', 'Description', 'X Coordinate','Y Coordinate','Community Area','IUCR']
    df.drop(redundant_columns, axis=1, inplace= True)
    
drop_redundant_columns(chicago_crime_df)

In [15]:
chicago_crime_df.head()

Case Number                    Date       Primary Type  \
0    HP485721  07/26/2008 02:30:00 PM    CRIMINAL DAMAGE   
1    JB372830  07/31/2018 10:57:00 AM  WEAPONS VIOLATION   
2    HN308568  04/27/2007 10:30:00 AM           BURGLARY   
3    JB361368  07/23/2018 08:55:00 AM            ASSAULT   
4    JE297624  07/11/2021 06:40:00 AM            ASSAULT   

                     Location Description  Arrest  Domestic  Beat District  \
0                                  STREET   False     False   423      4.0   
1                               POOL ROOM    True     False   413      4.0   
2                               RESIDENCE    True     False  1711     17.0   
3            NURSING HOME/RETIREMENT HOME   False     False  1115     11.0   
4  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  1231     12.0   

   Ward  Year   Latitude  Longitude  
0  10.0  2008  41.739980 -87.555120  
1   8.0  2018  41.726922 -87.599747  
2  39.0  2007  41.994138 -87.734959  
3  28.0  2018  41.881217 -87.730590  
4  27.0  2021  41.874174 -87.668082

In [16]:
chicago_crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2278726 entries, 0 to 2278725
Data columns (total 12 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Case Number           object 
 1   Date                  object 
 2   Primary Type          object 
 3   Location Description  object 
 4   Arrest                bool   
 5   Domestic              bool   
 6   Beat                  int64  
 7   District              string 
 8   Ward                  float64
 9   Year                  int64  
 10  Latitude              float64
 11  Longitude             float64
dtypes: bool(2), float64(3), int64(2), object(4), string(1)
memory usage: 178.2+ MB


In [ ]:
# A summary statistics of each column
chicago_crime_df.describe(include='all')

In [ ]:
chicago_crime_df.info()

__3. Handling missing values__

In [ ]:
# checking missing value count
chicago_crime_df.isnull().sum()

In [ ]:
# # Missing Values

# Function to calculate missing values by column
def missing_values_table(df):
        # Total missing values
        mis_val = df.isnull().sum()
        
        # Percentage of missing values
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        
        # Make a table with the results
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        
        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        
        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        
        # Print some summary information
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        
        # Return the dataframe with missing information
        return mis_val_table_ren_columns

In [ ]:
missing_values_table(chicago_crime_df)

- __Handling Missing Case Number__

We have one (1) missing Case Number in the entire dataset. We have decided to drop the row where Case Number is missing 

In [ ]:
# dropping the missing Case Number column null value
chicago_crime_df.dropna(subset=['Case Number'],inplace=True) # i chose to drop this because the percentage NA is insignificant

In [ ]:
chicago_crime_df.isna().sum()

- __Handling Missing values for the District Column__

We have just 12 missing value for the district column which is 0.0% of the entire dataset. We have decided to drop the row where location discription is missing.

In [ ]:
# dropping the missing district column null value
chicago_crime_df.dropna(subset=['District'],inplace=True) # i chose to drop this because the percentage NA is insignificant

In [ ]:
chicago_crime_df.District.unique()

In [ ]:
chicago_crime_df.isna().sum()

- __Handling Missing values for the ward Column (Using data imputation)__

In [ ]:
missing_values_table(chicago_crime_df)

In [ ]:
# checking the unique values of the ward column
chicago_crime_df.Ward.unique()

In [ ]:
chicago_crime_df.Ward

In [ ]:
# Using ffill(forward fill) to fill missing value in the ward column 
chicago_crime_df.Ward = chicago_crime_df.Ward.ffill() # Last observed value is carried forward

In [ ]:
chicago_crime_df.isna().sum()

- __Handling Missing values for the latitude and Longitude Column__ (Using Assosciation) <br>
    - The assumption here is that the latitude and longitude readings are points that show exact location, are closer to eachother within a district. if they are, we can group our data by districts and fill up missing Latitide and longitude values by the median or mode.

In [ ]:
chicago_crime_df.groupby(['District', 'Longitude', 'Latitude']).size()

There appears to be closeness in longitude and Latitude points when data is grouped based on districts, we can group based on district and fill up by taking the median.

In [ ]:
chicago_crime_df['Latitude'] = chicago_crime_df.groupby('District')['Latitude'].transform(lambda x: x.fillna(x.median()))
chicago_crime_df['Longitude'] = chicago_crime_df.groupby('District')['Longitude'].transform(lambda x: x.fillna(x.median()))

In [ ]:
chicago_crime_df.isna().sum()

- __Handling Missing values for the Location Description__

We have just 2877 missing value for the location description which is 0.1% of the entire dataset. We can decide to drop the row where location decription is missing or fill this row with the most crime location.

In [ ]:
chicago_crime_df.groupby(['District','Location Description']).size()

In [ ]:
# dropping the missing location description index row
chicago_crime_df.dropna(subset=['Location Description'],inplace=True) 

In [ ]:
chicago_crime_df.isna().sum()

# 4. EDA: Data Exploration

In [ ]:
# create a copy of the dataframe to avoid conflict
c_crime_df = chicago_crime_df.copy()

1. Total Number of cases

In [ ]:
# Number of Unique Case Number.

print(f"Out of {len(c_crime_df)}: {len(c_crime_df['Case Number'].unique())} are unique case numbers")  

Observation: Most case numbers are unique case numbers.

Number of Non-Unique cases

In [ ]:
# Out of all cases , how many are non_unique cases ? 

non_unique_case_number = c_crime_df[c_crime_df.duplicated('Case Number') | c_crime_df.duplicated('Case Number', keep=False)]['Case Number'].tolist() 

print(f"Length of non_unique: {len(non_unique_case_number)} out of {len(c_crime_df)}")

In [ ]:
non_unique_case_number_df = c_crime_df[c_crime_df.duplicated('Case Number', keep=False)]
non_unique_case_number_df.head() # Saving non_unique case number dataframe for future use.

Observation: There are 2275789 unique values of case number out of 2275836 , it means there are 2275789 cases registered with a single crime the rest 92 are the case numbers which have commited multiple crimes.

2. Top 10 - Number of cases each day

In [ ]:
# Function to convert Date from "DD/MM/YYYY HH:MM:SS" to "DD/MM/YYYY"

def date_column(df) : 
    date_list = []
    for i in df['Date'] : 
        date_spliter = i.split()[0]
        date_list.append(date_spliter)
    return date_list
date_column(c_crime_df)  


c_crime_df['date'] = date_column(c_crime_df) 

In [ ]:
# Top 10 - Number of cases each day? 

top_10_num_of_cases_each_day = c_crime_df.groupby(['date'])['Case Number'].count().reset_index(name = 'no_of_cases').set_index('no_of_cases').sort_index(ascending = False)[:10]
top_10_num_of_cases_each_day

In [ ]:
top_10_num_of_cases_each_day_df = pd.DataFrame(top_10_num_of_cases_each_day)
index_df = top_10_num_of_cases_each_day_df.index

sns.barplot(x = index_df , y = 'date' , data = top_10_num_of_cases_each_day_df)
plt.title("Most cases per day - Bar plot")
plt.xlabel("Number of Cases")
plt.ylabel("Date")
plt.show()

Observation: The average Number of cases per day could be more than 550

3. __Number of crime by type__</br>
We try to indentify the trend and highpoints or lowpoints.

In [ ]:
# Function to plot the serie value_counts
def plot_counts(serie, title):
    df = pd.DataFrame(serie.value_counts()[:15])
    df.columns = ["Freq"]
    df["Type"] = df.index
    fig = px.bar(df, y="Freq", x="Type", text="Freq", color="Freq", color_continuous_scale=px.colors.sequential.Blugrn)
    fig.update_traces(texttemplate="%{text:.2s}", textposition="outside")
    fig.update_layout(uniformtext_minsize=8, uniformtext_mode="hide")
    fig.update_layout(title_text=title)
    fig.show()


plot_counts(c_crime_df["Primary Type"], "Type of Crimes")

Observation: THEFT, BATTERY, CRIMINAL DAMAGE and NARCOTICS represents more the 65% of the crimes.

4. __Where does crimes happens the most?__

In [ ]:
plot_counts(c_crime_df["Location Description"], "Location of Crimes")

Crime happens mostly on the street

5. __Number of crimes per Year x Months__ </br>
Let's see the number of crimes per Year and Months...

In [ ]:
# first we convert the date column to datetime data type.

c_crime_df['Date'] = pd.to_datetime(chicago_crime_df['Date'],format = "%m/%d/%Y %I:%M:%S %p")
c_crime_df.set_index("Date", inplace=True)

In [ ]:
cases = pd.DataFrame(c_crime_df.resample("ME").size(), columns=["Number of cases"])
cases["Month"] = cases.index.month.astype(str)
cases["Year"] = cases.index.year.astype(str)
cases["Year-Month"] = cases["Year"].str.cat(cases["Month"].str.zfill(2), sep="-")
cases.head()

BY HEATMAP

In [ ]:
case_table = cases[["Month", "Year", "Number of cases"]]
case_table  = case_table.pivot(index='Month', columns='Year', values='Number of cases').fillna(0).astype(int)
case_table.index = case_table.index.astype(int)
case_table = case_table.sort_index()
case_table.index = ['Jan', 'Fev', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
plt.figure(figsize=(20, 10))
_ = sns.heatmap(
    case_table,
    annot=True,
    fmt="d",
    linewidths=0.5,
    cmap="Blues",
).set_title('Number of crimes in chicago per Year x Months')

The years between 2001 and 2012 were a tough time, the year 2008 was the worst with a mean of 71004 crimes per month, and the best year was the 2015 with a mean of 21916 crimes per month.

BY BAR

In [ ]:
fig = px.bar(
    cases,
    x="Year-Month",
    y="Number of cases",
    hover_data=["Year", "Month", "Number of cases"],
    color="Number of cases",
    height=600,
    width=1400,
    color_continuous_scale=px.colors.sequential.Blugrn

)
fig.update_layout(title_text='Crime counts per year-month in bars')
fig.show()

Similar graph but using bars to visualize though noisy but can be used. Again the 'tough time' during 2001 until 2012.

Are there specific location that crimes used to happen?

Using Map plot

In [ ]:
c_crime_df.info()

In [ ]:
map_marks = chicago_crime_df[['Latitude', 'Longitude']]
map_marks['Year'] =  pd.to_datetime(chicago_crime_df.Date, format="%m/%d/%Y %I:%M:%S %p").dt.year
map_marks.head()

In [ ]:
fig = px.scatter_mapbox(map_marks[map_marks.Year == 2016].sample(5000), lat="Latitude", lon="Longitude",
                        color_discrete_sequence=["fuchsia"], zoom=9, height=650)
fig.update_layout(mapbox_style="open-street-map")
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.show()

Look like there is crime everywhere in Chicago, in the centre, you can see more points but also is a crowded place.

What is the arrested rate per crime?

In [ ]:
cases = c_crime_df.groupby('Primary Type')['Arrest'].mean().reset_index()
cases['Arrest (%)'] = cases['Arrest'].round(4)*100
cases.sort_values('Arrest', inplace=True)
fig = px.bar(
    cases,
    orientation='h',
    x="Arrest (%)",
    y="Primary Type",
    color="Arrest (%)",
    text="Arrest (%)",
    height=800,
    #width=800,
        color_continuous_scale=px.colors.sequential.Blugrn

)
fig.update_traces(texttemplate="%{text:.4s}", textposition="outside")
fig.update_layout(title_text="Arrested Rate in % per Type Crime in Chicago.")

fig.show()

The “VIOLATION” crimes look strongly correlated to the arrested rate once the 'cop' must be present in the location to declare the violation.

In [ ]:
chicago_crime_df.info()

In [ ]:
# Try to find the correlation of the columns 

# Select only numeric columns
numeric_columns = chicago_crime_df.select_dtypes(include=['number'])

# Calculate correlation
numeric_columns.corr()

In [ ]:
#plot a heatmap showing correlation
heatmap = sns.heatmap(numeric_columns.corr(), vmin=-1, annot=True)
heatmap.set_title('Chicago Crime Correlation', fontdict={'fontsize':14}, pad=10)

**_Observation_**
1. The Beat and Ward are positively correlated.

In [ ]:
chicago_crime_df.groupby(['Primary Type', 'Arrest','Domestic']).size().sort_values(ascending=False).reset_index().rename(columns={0:'Crime Count'})

In [ ]:
plt.figure(figsize=(12, 6))
sns.countplot(x='Primary Type', hue='Arrest', data=chicago_crime_df[:8])

# 5. Feature Engineering

Create new columns for Months, Day, and Season

__Creating month column__

In [ ]:
# Create new columns for Months
def extract_and_map_month(date):
    # Ensure Date column have been converted to datetime format
    month_names = chicago_crime_df['Date'].dt.month_name()
# dt is an attribute of the date column object
# month_name() is a method
    return month_names

In [ ]:
chicago_crime_df['Month'] = extract_and_map_month(chicago_crime_df['Date'])

# Group by month and count number of crimes
monthly_crime_counts = chicago_crime_df.groupby('Month').size()

In [ ]:
# Monthly plot for crimes
plt.figure(figsize=(10, 6))
monthly_crime_counts.plot(kind='bar', color='green')
plt.title('Monthly Crime Counts in Chicago')
plt.xlabel('Month')
plt.ylabel('Number of Crimes')
plt.xticks(rotation=0)
plt.show()

__Creating Day Column__

In [ ]:
# Create new columns for Day
def extract_and_map_month(date):
    # ensure the input is in datetime format
    day_names = chicago_crime_df['Date'].dt.day_name()
# dt is an attribute of the date column object
# month_name() is a method
    return day_names

In [ ]:
chicago_crime_df['Day'] = extract_and_map_month(chicago_crime_df['Date'])

# Group by day and count number of crimes
daily_crime_counts = chicago_crime_df.groupby('Day').size()


In [ ]:
# daily plot for crimes
plt.figure(figsize=(12, 6))
daily_crime_counts.plot(kind='line', color='green', marker='o', markersize=10)
plt.title('Daily Crime Counts in Chicago')
plt.xlabel('Day of the Month')
plt.ylabel('Number of Crimes')
plt.grid(True)
plt.show()

__Creating Seasons column__

In [ ]:
# Function to define various month and their corresponding season
def get_season(month):
    if month in ['December','January','February']:
        return 'Winter'
    elif month in ['March','April','May']:
        return 'Spring'
    elif month in ['June','July','August']:
        return 'Summer'
    else:
        return 'Fall'

In [ ]:
chicago_crime_df['Season'] = chicago_crime_df['Month'].apply(get_season)

# Group by season and count number of crimes
seasonal_crime_counts = chicago_crime_df.groupby('Season').size()


In [ ]:
# seasonal plot for crimes
plt.figure(figsize=(10, 6))
seasonal_crime_counts.plot(kind='bar', color='green')
plt.title('Seasonal Crime Counts in Chicago')
plt.xlabel('Season')
plt.ylabel('Number of Crimes')
plt.xticks(rotation=0)
plt.show()

# 6. Frequency Distribution of Crimes

 __Monthly Crime Frequency__

In [ ]:
# created another df where the Date would be the Index
crimes_df = chicago_crime_df.set_index(['Date'])

In [ ]:
# show a trend analysis of arrest over years, months and days
plt.subplot()
#plt.figure(figsize=(8, 6))

# this gives us data where arrest was made
arrests = crimes_df[crimes_df['Arrest'] == True]['Arrest']

#yearly arrest 
arrests.resample('YE').count().plot() #the 'A' is an offset Data that indicates annual, view documentation for further understanding
plt.title('Yearly arrests')
plt.show()

# quarter start arrest
arrests.resample('QS').count().plot() 
plt.title('Quarterly Arrests')
plt.show()

# monthly arrest
arrests.resample('ME').count().plot() 
plt.title('Monthly Arrests')
plt.show()



__Observation__
1. There is a downward trend of arrest over the years from 2008 upwards this is evident even in the quarterly arrests.
2. 2020 recorded the least monthly arrest.

In [ ]:
crime_count_by_month = chicago_crime_df.groupby('Month')['Arrest'].count()
print(crime_count_by_month)

__Seasonal Crime Frequency__

In [ ]:
crime_count_by_season = chicago_crime_df.groupby('Season')['Season'].count()
crime_count_by_season

In [ ]:
# Group by season and count number of arrests
seasonal_arrests = chicago_crime_df[chicago_crime_df['Arrest'] == True]['Season'].value_counts()

# Plot pie chart for seasonal arrests
plt.figure(figsize=(8, 6))
seasonal_arrests.plot(kind='pie', autopct='%1.1f%%', colors=['skyblue', 'lightgreen', 'lightcoral', 'orange'])
plt.title('Seasonal Arrests in Chicago')
plt.ylabel('')
plt.show()

 __Daily Crime Frequency__

In [ ]:
crime_count_per_day = chicago_crime_df.groupby('Day')['Day'].count()
print(crime_count_per_day)

In [ ]:
# This will suppress any warnings related to the change in figure layout.
import warnings
warnings.filterwarnings("ignore", message="The figure layout has changed to tight")

# Frequency of arrests for each day
sns.catplot(
    data=chicago_crime_df, y="Day", hue="Arrest", kind="count",
    palette="pastel", legend=True, edgecolor=".6",
)

In [ ]:
#Location with the most crime

chicago_crime_df.groupby('Location Description')['Primary Type'].size().sort_values(ascending=False).head(20)

# 7. Visualize Data Distribution

In [ ]:
# Visualize the distribution of crime types
plt.figure(figsize=(12, 6))
sns.countplot(data=chicago_crime_df, x='Primary Type')
plt.xticks(rotation=90)
plt.title('Distribution of Crime Types')

# 8. Investigate Relationships

In [ ]:
numeric_columns.corr()

In [ ]:
heatmap = sns.heatmap(numeric_columns.corr(), vmin=-1, annot=True)
heatmap.set_title('Chicago Crime Correlation', fontdict={'fontsize':14}, pad=10)

# i can infer the relationship between the districts and the ward

In [ ]:
crimes_df.head()

In [ ]:
# Investigate Relationships
# Explore the relationship between Location description and crime type

# Create a cross-tabulation of crime location and crime type
crime_location_type = pd.crosstab(chicago_crime_df['Location Description'], chicago_crime_df['Primary Type'])

# Plot heatmap
plt.figure(figsize=(18, 12))
sns.heatmap(crime_location_type, cmap='YlGnBu', linewidths=.1)
plt.title('Frequency of Crime Types at Different Locations')
plt.xlabel('Crime Type')
plt.ylabel('Location Description')
plt.xticks(rotation=90)
plt.yticks(rotation=0)
plt.show()

# 8. Examine Crime Locations

In [ ]:
# Location description where crime happens the most
most_common_crime_location = chicago_crime_df['Location Description'].value_counts().idxmax()

In [ ]:
most_common_crime_location

# 9. Identify Hotspots

In [ ]:
# using district to determine areas with higest crime rate(True/False)

# Calculate total number of arrests for each district
district_arrests = chicago_crime_df.groupby('District')['Arrest'].sum()
# print(district_arrests)

# Determine the districts with the highest arrest rates
highest_arrest_rate_districts = district_arrests.nlargest(5)  # Get the top 5 districts with highest arrest rates

# print(highest_arrest_rate_districts)

# Create a boolean column indicating whether each district has one of the highest arrest rates
chicago_crime_df['Highest_Arrest_Rate'] = chicago_crime_df['District'].isin(highest_arrest_rate_districts.index)

# Print the result (True/False) for each district
print(chicago_crime_df[['District', 'Highest_Arrest_Rate']].drop_duplicates())


# using district to determine areas with higest crime rate(True)


# 11. Compare Crime Rates

In [ ]:
# mapping crimes based on locations

# Filter out rows with missing latitude or longitude values
chicago_crime_df = chicago_crime_df.dropna(subset=['Latitude', 'Longitude'])

# Create a map centered around the mean latitude and longitude
crime_map = folium.Map(location=[chicago_crime_df['Latitude'].mean(), chicago_crime_df['Longitude'].mean()], zoom_start=11)

# Add markers for each crime location
for index, row in chicago_crime_df.iterrows():
    folium.Marker(location=[row['Latitude'], row['Longitude']], popup=row['Primary Type']).add_to(crime_map)

# Save the map to an HTML file
crime_map.save('crime_map.html')

# 12. Summarize and Visualize Insights